apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs

```diff
@@ -46,6 +116,159 @@ private SpannerConversionOptions(bool useDBNull)
         internal static SpannerConversionOptions ForConnectionStringBuilder(SpannerConnectionStringBuilder builder) =>
-            builder == null ? Default : builder.UseClrDefaultForNull ? s_useClrDefaultForNull : s_useDBNullForNull;
+            builder == null ? Default : builder.ConversionOptions;
+
+        internal SpannerConversionOptions Clone() => new SpannerConversionOptions(UseDBNull)
+        {
+            DateTimeToConfiguredSpannerType = DateTimeToConfiguredSpannerType,
+            DecimalToConfiguredSpannerType = DecimalToConfiguredSpannerType,
+            DateToConfiguredClrType = DateToConfiguredClrType,
+            Float64ToConfiguredClrType = Float64ToConfiguredClrType
+        };
+
+        internal SpannerConversionOptions WithClrDefaultForNullSetting(bool useClrDefaultToNull)
+        {
+            var clone = Clone();
+            clone.UseDBNull = !useClrDefaultToNull;
+            return clone;
+        }
+
+        internal SpannerConversionOptions WithClrToSpannerMappings(string clrToSpannerMappings)
+        {
+            var clone = Clone();
+            ValidateAndParseClrToSpannerTypeMappings(clrToSpannerMappings, clone);
+            return clone;
+        }
+
+        internal SpannerConversionOptions WithSpannerToClrMappings(string spannerToClrMappings)
+        {
+            var clone = Clone();
+            ValidateAndParseSpannerToClrTypeMappings(spannerToClrMappings, clone);
+            return clone;
+        }
+
+        private void ValidateAndParseSpannerToClrTypeMappings(string input, SpannerConversionOptions options)
+        {
+            // Empty string or null is valid to unset the values.
+            if (string.IsNullOrEmpty(input))
+            {
+                // Set defaults.
+                options.SetSpannerToClrTypeDefaults();
+                return;
+            }
+
+            // Get all type mappings.
+            var mappings = input.Split(new char[] { ',', ';' }, StringSplitOptions.RemoveEmptyEntries);
+
+            // Validate that mappings are valid.
+            if (mappings == null || !mappings.Any())
+            {
+                throw new ArgumentException($"'{input}' is not a valid value for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // Validate that mapping matches one of the pre-defined values.
+            var invalidMappings = mappings.Except(AllowedSpannerToClrTypeMappings, StringComparer.OrdinalIgnoreCase);
+            if (invalidMappings.Any())
+            {
+                throw new ArgumentException($"The following mappings are invalid:'{string.Join(",", invalidMappings)}' for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // Check multiples for each type.
+            // Currently, we have Float64 and Date to check.
+            if (mappings.Count(j => j.StartsWith("Float64", StringComparison.OrdinalIgnoreCase)) > 1)
```

Comment:
I think we might want to have a more "data-oriented" approach to this, grouping options by affected type - but we don't need to do that immediately. (I'm happy to send a refactoring PR.)

Reply:
I would probably do something like having a `Dictionary<string, (SpannerDbType, Type)>` for each option to "source and target types". Then you can find out any options that conflict by whether there multiple options with the same source. (That can also be a case-insensitive dictionary.) I can picture the code, but it's somewhat fiddly to actually describe. But basically we'd end up with the two dictionaries (one in each direction) having almost all the type-specific aspects, so adding a new conversion option should be a small change.